File: python/mirascope/ops/_internal/configuration.py
"""Configuration utilities for Mirascope ops module initialization and setup."""

from __future__ import annotations

import logging
from collections.abc import Iterator
from contextlib import contextmanager
from typing import TYPE_CHECKING

from opentelemetry import trace as otel_trace
from opentelemetry.sdk.trace import TracerProvider

if TYPE_CHECKING:
    from opentelemetry.trace import Tracer

DEFAULT_TRACER_NAME = "mirascope.llm"

logger = logging.getLogger(__name__)

_tracer_provider: TracerProvider | None = None
_tracer_name: str = DEFAULT_TRACER_NAME
_tracer_version: str | None = None
_tracer: Tracer | None = None


def configure(
    *,
    tracer_provider: TracerProvider,
    tracer_name: str = DEFAULT_TRACER_NAME,
    tracer_version: str | None = None,
) -> None:
    """Configure the ops module for tracing.

    Args:
        tracer_provider: The TracerProvider to use for tracing.
        tracer_name: Tracer name to use when creating a tracer.
            Defaults to "mirascope.llm".
        tracer_version: Optional tracer version.

    Example:
        ```python
        from mirascope import ops
        from opentelemetry.sdk.trace import TracerProvider

        provider = TracerProvider()
        ops.configure(tracer_provider=provider)
        ```
    """
    global _tracer_provider, _tracer_name, _tracer_version, _tracer

    _tracer_provider = tracer_provider
    otel_trace.set_tracer_provider(tracer_provider)

    _tracer_name = tracer_name
    _tracer_version = tracer_version

    _tracer = tracer_provider.get_tracer(_tracer_name, _tracer_version)


def set_tracer(tracer: Tracer | None) -> None:
    """Set the configured tracer instance."""
    global _tracer
    _tracer = tracer


def get_tracer() -> Tracer | None:
    """Return the configured tracer instance."""
    return _tracer


@contextmanager
def tracer_context(tracer: Tracer | None) -> Iterator[Tracer | None]:
    """Context manager for temporarily setting a tracer.

    Temporarily sets the tracer for the duration of the context and restores
    the previous tracer when the context exits.

    Args:
        tracer: The tracer to use within the context.

    Yields:
        The tracer that was set.

    Example:
        ```python
        from mirascope import ops
        from opentelemetry.sdk.trace import TracerProvider

        provider = TracerProvider()
        tracer = provider.get_tracer("my-tracer")

        with ops.tracer_context(tracer):
            # Use the tracer within this context
            ...
        # Previous tracer is restored here
        ```
    """
    previous_tracer = get_tracer()
    set_tracer(tracer)
    try:
        yield tracer
    finally:
        set_tracer(previous_tracer)

File: typescript/src/ops/_internal/configuration.ts
/**
 * Configuration utilities for Mirascope ops module initialization and setup.
 */

import type { Tracer } from "@opentelemetry/api";
import type { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";

const DEFAULT_TRACER_NAME = "mirascope.llm";

let _tracerProvider: NodeTracerProvider | null = null;
let _tracerName: string = DEFAULT_TRACER_NAME;
let _tracerVersion: string | undefined = undefined;
let _tracer: Tracer | null = null;

/**
 * Options for configuring the ops module.
 */
export interface ConfigureOptions {
  /** Custom tracer provider */
  tracerProvider: NodeTracerProvider;
  /** Tracer name (default: "mirascope.llm") */
  tracerName?: string;
  /** Tracer version */
  tracerVersion?: string;
}

/**
 * Configure the ops module for tracing.
 *
 * Requires a tracer provider to be passed in.
 *
 * @example
 * With custom tracer provider:
 * ```typescript
 * import { ops } from 'mirascope';
 * import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
 *
 * const provider = new NodeTracerProvider();
 * ops.configure({ tracerProvider: provider });
 * ```
 */
export function configure(options: ConfigureOptions): void {
  const {
    tracerProvider,
    tracerName = DEFAULT_TRACER_NAME,
    tracerVersion,
  } = options;

  _tracerProvider = tracerProvider;

  // Register the provider which sets up both tracer provider AND context manager
  // This is critical for async context propagation (span nesting)
  _tracerProvider.register();

  _tracerName = tracerName;
  _tracerVersion = tracerVersion;
  _tracer = _tracerProvider.getTracer(_tracerName, _tracerVersion);
}

/**
 * Get the configured tracer instance.
 *
 * @returns The current tracer, or null if not configured.
 */
export function getTracer(): Tracer | null {
  return _tracer;
}

/**
 * Set the tracer instance.
 *
 * @param tracer - The tracer to set, or null to clear.
 */
export function setTracer(tracer: Tracer | null): void {
  _tracer = tracer;
}

/**
 * Run a function with a temporary tracer.
 *
 * Temporarily sets the tracer for the duration of the function execution
 * and restores the previous tracer when the function returns.
 *
 * @param tracer - The tracer to use during function execution.
 * @param fn - The function to execute.
 * @returns The result of the function.
 *
 * @example
 * ```typescript
 * import { ops } from 'mirascope';
 * import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
 *
 * const provider = new NodeTracerProvider();
 * const tracer = provider.getTracer('my-tracer');
 *
 * const result = ops.tracerContext(tracer, () => {
 *   // Use the tracer within this context
 *   return someOperation();
 * });
 * // Previous tracer is restored here
 * ```
 */
export function tracerContext<T>(tracer: Tracer | null, fn: () => T): T {
  const previous = _tracer;
  _tracer = tracer;
  try {
    return fn();
  } finally {
    _tracer = previous;
  }
}

/**
 * Force flush all pending spans to be exported.
 *
 * Convenience wrapper around the standard OpenTelemetry
 * {@link NodeTracerProvider.forceFlush} method. Delegates directly to the
 * underlying provider; no-ops when {@link configure} has not been called.
 *
 * @returns A promise that resolves when all spans have been flushed.
 *
 * @example
 * ```typescript
 * import { ops } from 'mirascope';
 * import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
 *
 * const provider = new NodeTracerProvider();
 * ops.configure({ tracerProvider: provider });
 * ops.instrumentLLM();
 *
 * // Run your traced operations
 * await myTracedFunction();
 *
 * // Force flush before script exits
 * await ops.forceFlush();
 * ```
 */
export async function forceFlush(): Promise<void> {
  if (_tracerProvider) {
    await _tracerProvider.forceFlush();
  }
}

/**
 * Shutdown the tracer provider and flush all pending spans.
 *
 * Convenience wrapper around the standard OpenTelemetry
 * {@link NodeTracerProvider.shutdown} method. Delegates directly to the
 * underlying provider; no-ops when {@link configure} has not been called.
 *
 * @returns A promise that resolves when the provider has been shut down.
 *
 * @example
 * ```typescript
 * import { ops } from 'mirascope';
 *
 * // At application shutdown
 * process.on('SIGTERM', async () => {
 *   await ops.shutdown();
 *   process.exit(0);
 * });
 * ```
 */
export async function shutdown(): Promise<void> {
  if (_tracerProvider) {
    await _tracerProvider.shutdown();
  }
}

/**
 * Reset configuration state.
 *
 * This is primarily useful for testing to ensure a clean state between tests.
 */
export function resetConfiguration(): void {
  _tracerProvider = null;
  _tracerName = DEFAULT_TRACER_NAME;
  _tracerVersion = undefined;
  _tracer = null;
}
